File: stage2-parser/src/grammar/Parser.java
package grammar;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.LinkedHashSet;

import errors.SyntaxError;
import lexer.Lexer;
import lexer.Tag;
import lexer.Token;

public class Parser {
	private Lexer lexer;
	private Token token;
	private LinkedHashSet<Integer> firstPrimaryExpression = new LinkedHashSet<Integer>();
	private LinkedHashSet<Integer> firstUnaryExpression = new LinkedHashSet<Integer>();
	private LinkedHashSet<Integer> firstExtendedMultiplicativeExpression = new LinkedHashSet<Integer>();
	private LinkedHashSet<Integer> firstMultiplicativeExpression = new LinkedHashSet<Integer>();
	private LinkedHashSet<Integer> firstExtendedAdditiveExpression = new LinkedHashSet<Integer>();
	private LinkedHashSet<Integer> firstAdditiveExpression = new LinkedHashSet<Integer>();
	
	public Parser (FileInputStream input) {
		lexer = new Lexer(input);
		
		firstPrimaryExpression.add(Tag.ID); firstPrimaryExpression.add(Tag.NUMBER);
		firstPrimaryExpression.add(Tag.TRUE); firstPrimaryExpression.add(Tag.FALSE);
		firstPrimaryExpression.add((int) '(');
		
		firstUnaryExpression.add((int) '-'); firstUnaryExpression.add((int) '!');
		firstUnaryExpression.addAll(firstPrimaryExpression);
		
		firstExtendedMultiplicativeExpression.add((int) '*'); firstExtendedMultiplicativeExpression.add((int) '/');
		firstExtendedMultiplicativeExpression.add(Tag.MOD);
		
		firstMultiplicativeExpression.containsAll(firstUnaryExpression);
		
		firstExtendedAdditiveExpression.add((int) '+'); firstExtendedAdditiveExpression.add((int) '-');
		
		firstAdditiveExpression.addAll(firstMultiplicativeExpression);
	}

	private void check(int tag) throws SyntaxError, IOException {
		if (token.getTag() == tag) {
			token = lexer.scan();
		} else {
			throw new SyntaxError();
		}
	}

	public void analyze() throws SyntaxError, IOException {
		token = lexer.scan();
		//program();
	}
	
	private void primaryExpression() throws SyntaxError, IOException {
		if (firstPrimaryExpression.contains(token.getTag())) {
			switch(token.getTag()) {
			case Tag.ID : check(Tag.ID); break;
			case Tag.NUMBER : check(Tag.NUMBER); break;
			case Tag.TRUE : check(Tag.TRUE); break;
			case Tag.FALSE : check(Tag.FALSE); break;
			case ((int) '(') :
				check((int) '(');
				//expression();
				check((int) ')');
				break;
			}
		} else {
			throw new SyntaxError();
		}
	}
	
	private void unaryExpression() throws SyntaxError, IOException {
		if (firstUnaryExpression.contains(token.getTag())) {
			switch(token.getTag()) {
			case ((int) '-') : check((int) '-'); unaryExpression(); break;
			case ((int) '!') : check((int) '!'); unaryExpression(); break;
			default			 : primaryExpression(); break;
			}
		} else {
			throw new SyntaxError();
		}
	}
	
	private void extendedMultiplicativeExpression() throws SyntaxError, IOException {
		if (firstExtendedMultiplicativeExpression.contains(token.getTag())) {
			switch(token.getTag()) {
			case ((int) '*') : 
				check((int) '*');
				unaryExpression();
				extendedMultiplicativeExpression();
				break;
			case ((int) '/') : 
				check((int) '/');
				unaryExpression();
				extendedMultiplicativeExpression();
				break;
			case Tag.MOD : 
				check(Tag.MOD);
				unaryExpression();
				extendedMultiplicativeExpression();
				break;
			}
		} else {
			// do nothing
		}
	}
	
	protected void multiplicativeExpression() throws SyntaxError, IOException {
		if (firstMultiplicativeExpression.contains(token.getTag())) {
			unaryExpression();
			extendedMultiplicativeExpression();
		}
	}
	
	protected void extendedAdditiveExpression() throws SyntaxError, IOException {
		if (firstExtendedAdditiveExpression.contains(token.getTag())) {
			switch(token.getTag()) {
			case ((int) '+') : 
				check((int) '+');
				multiplicativeExpression();
				extendedAdditiveExpression();
				break;
			case ((int) '-') : 
				check((int) '-');
				multiplicativeExpression();
				extendedAdditiveExpression();
				break;
			}
		} else {
			// do nothing
		}
	}
	
	protected void additiveExpression() throws SyntaxError, IOException {
		if (firstAdditiveExpression.contains(token.getTag())) {
			multiplicativeExpression();
			extendedAdditiveExpression();
		}
	}
}

File: stage-2/Parser.py
from Lexer import *

class Parser:
	__lex = None
	__token = None

	def __init__(self, filepath):
		self.__lex = Lexer(filepath)
		self.__token = None

		self.__firstPrimaryExpression = set((Tag.ID, Tag.NUMBER, Tag.TRUE, Tag.FALSE, ord('(')))

		self.__firstUnaryExpression = self.__firstPrimaryExpression.union( set((ord('-'), ord('!'))) )
		
		self.__firstExtendedMultiplicativeExpression = set((ord('*'), ord('/'), Tag.MOD))

		self.__firstMultiplicativeExpression = self.__firstUnaryExpression

	def error(self, extra = None):
		text = 'Line ' + str(self.__lex.getLine()) + " - " 
		if extra == None:
			text = text + "."
		else:
			text = text + extra
		raise Exception(text)

	def __check(self, tag):
		if self.__token.getTag() == tag:
			self.__token = self.__lex.scan()
		else:
			text = 'Line ' + str(self.__lex.getLine()) + " - expected "
			if tag != Tag.ID:
				text = text + str(Token(tag)) + " before " + str(self.__token) 
			else:
				text = text + "an identifier before " + str(self.__token) 
			raise Exception(text)
	
	def analize(self):
		self.__token = self.__lex.scan()
		## self.__program()

	def __primaryExpression(self):
		if self.__token.getTag() in self.__firstPrimaryExpression:
			if self.__token.getTag() == Tag.ID:
				self.__check(Tag.ID)
			elif self.__token.getTag() == Tag.NUMBER:
				self.__check(Tag.NUMBER)
			elif self.__token.getTag() == Tag.TRUE:
				self.__check(Tag.TRUE)
			elif self.__token.getTag() == Tag.FALSE:
				self.__check(Tag.FALSE)
			elif self.__token.getTag() == ord('('):
				self.__check('(')
				# self.__expression()
				self.__check(')')
		else:
			self.error('Syntax Error')
		
	def __unaryExpression(self):
		if self.__token.getTag() in self.__firstPrimaryExpression:
			if self.__token.getTag() == ord('-'):
				self.__check(ord('-'))
				self.__unaryExpression()
			elif self.__token.getTag() == ord('!'):
				self.__check(ord('!'))
				self.__unaryExpression
			else:
				self.__primaryExpression()
		else: 
			self.error('Syntax Error')
		
	def __extendedMultiplicativeExpression(self):
		if self.__token.getTag() in self.__firstExtendedMultiplicativeExpression:
			if self.__token.getTag() == ord('*'):
				self.__check(ord('*'))
				self.__check(ord('*'))
				self.__unaryExpression()
				self.__extendedMultiplicativeExpression()
			elif self.__token.getTag() == ord('/'):
				self.__check(ord('/'))
				self.__unaryExpression()
				self.__extendedMultiplicativeExpression()
			elif self.__token.getTag() == Tag.MOD:
				self.__check(Tag.MOD)
				self.__unaryExpression()
				self.__extendedMultiplicativeExpression()